File: packages/failsafe-core/src/scoring.ts
import type { FailureReport, RiskFactor } from './types';
import { SEVERITY_WEIGHTS } from './taxonomy';
import { normalizeScore } from './utils';

export function calculateImpactScore(report: FailureReport): number {
  // Scale severity to leave room for domain/status bonuses
  let score = SEVERITY_WEIGHTS[report.severity] * 0.7;

  if (report.domain === 'security') {
    score += 20;
  }
  if (report.domain === 'agent-orchestration') {
    score += 12;
  }

  if (report.verificationStatus !== 'resolved') {
    score += 8;
  }

  if (report.recommendations && report.recommendations.length > 2) {
    score += 5;
  }

  return normalizeScore(Math.round(score));
}

export function calculateUrgencyScore(report: FailureReport): number {
  const severityScore = SEVERITY_WEIGHTS[report.severity];
  const isUnverified = report.verificationStatus === 'unverified' ? 20 : 0;
  const isCritical = report.severity === 'critical' ? 30 : 0;

  return normalizeScore(severityScore + isUnverified + isCritical);
}

export function createRiskFactorsFromReport(report: FailureReport): RiskFactor[] {
  const factors: RiskFactor[] = [];

  factors.push({
    name: 'severity',
    weight: 0.3,
    value: SEVERITY_WEIGHTS[report.severity],
  });

  const verificationScore = report.verificationStatus === 'unverified' ? 80 : report.verificationStatus === 'verified' ? 40 : 20;
  factors.push({
    name: 'verification_status',
    weight: 0.2,
    value: verificationScore,
  });

  const domainScore = report.domain === 'security' ? 90 : report.domain === 'agent-orchestration' ? 75 : 60;
  factors.push({
    name: 'domain_criticality',
    weight: 0.25,
    value: domainScore,
  });

  const resolutionScore = report.verificationStatus === 'resolved' ? 10 : 70;
  factors.push({
    name: 'resolution_status',
    weight: 0.25,
    value: resolutionScore,
  });

  return factors;
}

export function calculateCompositeScore(report: FailureReport): number {
  const impact = calculateImpactScore(report) * 0.4;
  const urgency = calculateUrgencyScore(report) * 0.35;
  const baseline = 25;

  return normalizeScore(impact + urgency + baseline);
}

export function rankReportsByScore(reports: FailureReport[]): Array<{ report: FailureReport; score: number }> {
  return reports
    .map((report) => ({
      report,
      score: calculateCompositeScore(report),
    }))
    .sort((a, b) => b.score - a.score);
}

File: packages/failsafe-py/failsafe/scoring.py
"""FailSafe Risk Scoring"""

from __future__ import annotations

from typing import List, Tuple

from .types import FailureReport, RiskFactor
from .taxonomy import SEVERITY_WEIGHTS
from .utils import normalize_score


def calculate_impact_score(report: FailureReport) -> float:
    score = SEVERITY_WEIGHTS.get(report.severity, 0) * 0.7
    if report.domain == "security":
        score += 20
    if report.domain == "agent-orchestration":
        score += 12
    if report.verification_status != "resolved":
        score += 8
    if report.recommendations and len(report.recommendations) > 2:
        score += 5
    return normalize_score(round(score))


def calculate_urgency_score(report: FailureReport) -> float:
    severity_score = SEVERITY_WEIGHTS.get(report.severity, 0)
    is_unverified = 20 if report.verification_status == "unverified" else 0
    is_critical_sev = 30 if report.severity == "critical" else 0
    return normalize_score(severity_score + is_unverified + is_critical_sev)


def create_risk_factors_from_report(report: FailureReport) -> List[RiskFactor]:
    factors: List[RiskFactor] = []

    factors.append(RiskFactor(
        name="severity", weight=0.3,
        value=SEVERITY_WEIGHTS.get(report.severity, 0),
    ))

    verification_score = (
        80 if report.verification_status == "unverified"
        else 40 if report.verification_status == "verified"
        else 20
    )
    factors.append(RiskFactor(
        name="verification_status", weight=0.2, value=verification_score,
    ))

    domain_score = (
        90 if report.domain == "security"
        else 75 if report.domain == "agent-orchestration"
        else 60
    )
    factors.append(RiskFactor(
        name="domain_criticality", weight=0.25, value=domain_score,
    ))

    resolution_score = 10 if report.verification_status == "resolved" else 70
    factors.append(RiskFactor(
        name="resolution_status", weight=0.25, value=resolution_score,
    ))

    return factors


def calculate_composite_score(report: FailureReport) -> float:
    impact = calculate_impact_score(report) * 0.4
    urgency = calculate_urgency_score(report) * 0.35
    baseline = 25
    return normalize_score(impact + urgency + baseline)


def rank_reports_by_score(
    reports: List[FailureReport],
) -> List[Tuple[FailureReport, float]]:
    scored = [(r, calculate_composite_score(r)) for r in reports]
    scored.sort(key=lambda x: x[1], reverse=True)
    return scored
